cpp/src/traversal/extract_bfs_paths_impl.cuh

```diff
@@ -0,0 +1,378 @@
+/*
+ * Copyright (c) 2021, NVIDIA CORPORATION.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+#pragma once
+
+#include <utilities/graph_utils.cuh>
+
+#include <cugraph/algorithms.hpp>
+#include <cugraph/detail/graph_utils.cuh>
+#include <cugraph/detail/utility_wrappers.hpp>
+#include <cugraph/graph_view.hpp>
+#include <cugraph/utilities/error.hpp>
+#include <cugraph/utilities/host_scalar_comm.cuh>
+#include <cugraph/utilities/shuffle_comm.cuh>
+#include <cugraph/vertex_partition_device_view.cuh>
+
+#include <raft/handle.hpp>
+#include <rmm/exec_policy.hpp>
+
+#include <thrust/binary_search.h>
+#include <thrust/fill.h>
+#include <thrust/iterator/counting_iterator.h>
+#include <thrust/iterator/zip_iterator.h>
+#include <thrust/transform.h>
+#include <thrust/tuple.h>
+
+namespace cugraph {
+
+namespace detail {
+
+template <typename vertex_t>
+struct mg_partition_vertices {
+  int rank_;
+  vertex_t const* d_partition_offsets_begin_;
+  vertex_t const* d_partition_offsets_end_;
+
+  int __device__ operator()(vertex_t v) const
+  {
+    return thrust::distance(
+             d_partition_offsets_begin_,
+             thrust::upper_bound(
+               thrust::seq, d_partition_offsets_begin_, d_partition_offsets_end_, v)) -
+           1;
+  }
+};
+
+template <typename vertex_t>
+struct map_index_to_path_offset {
+  size_t max_path_length_;
+  vertex_t const* distances_;
+  vertex_t const* destinations_;
+
+  size_t __device__ operator()(vertex_t idx)
+  {
+    size_t tmp = idx;
+    return (tmp * max_path_length_) + distances_[destinations_[idx]];
+  }
+};
+
+template <typename vertex_t>
+struct update_paths {
+  vertex_t* paths_;
+  vertex_t invalid_vertex_;
+
+  void __device__ operator()(thrust::tuple<vertex_t, size_t> tuple)
+  {
+    auto next_v = thrust::get<0>(tuple);
+    auto offset = thrust::get<1>(tuple);
+
+    if (next_v != invalid_vertex_) paths_[offset] = next_v;
+  }
+};
+
+template <typename vertex_t>
+struct mg_lookup_distance {
+  vertex_t v_local_first_;
+  vertex_t const* distances_;
+
+  thrust::tuple<vertex_t, vertex_t, int> __device__
+  operator()(thrust::tuple<vertex_t, vertex_t, int> const& tuple)
```

Comment:
What's the third tuple element (type int)?

Reply:
The original rank.  Each rank has to retrieve results from another rank.  The current implementation is:
1. Construct the tuple (current_vertex, offset_in_current_frontier_array my_rank)
2. Shuffle the tuples based on which rank contains the predecessor information for current_vertex
3. Replace current_vertex with previous_vertex
4. Shuffle the tuples back to the original rank where the data is required
5. Update current_frontier in the proper location with the predecessor value

I could achieve the same result without the int parameter by writing a more sophisticated communication pattern (sort the vertices by which rank they are assigned to, do block transfers keeping track of the offsets for each rank, etc).  I suspect this would be more time efficient as well as memory efficient.  But I was trying to reuse existing functions as much as possible to get something working.  We can optimize this more later.

Reply:
Ah... OK, so I guess this code can go away if we implement collect_values_for_vertices.